File: packages/react-native-audio-api/src/core/AudioBuffer.ts
import { IAudioBuffer } from '../interfaces';
import { IndexSizeError } from '../errors';

export default class AudioBuffer {
  readonly length: number;
  readonly duration: number;
  readonly sampleRate: number;
  readonly numberOfChannels: number;
  /** @internal */
  public readonly buffer: IAudioBuffer;

  constructor(buffer: IAudioBuffer) {
    this.buffer = buffer;
    this.length = buffer.length;
    this.duration = buffer.duration;
    this.sampleRate = buffer.sampleRate;
    this.numberOfChannels = buffer.numberOfChannels;
  }

  public getChannelData(channel: number): Float32Array {
    if (channel < 0 || channel >= this.numberOfChannels) {
      throw new IndexSizeError(
        `The channel number provided (${channel}) is outside the range [0, ${this.numberOfChannels - 1}]`
      );
    }
    return this.buffer.getChannelData(channel);
  }

  public copyFromChannel(
    destination: Float32Array,
    channelNumber: number,
    startInChannel: number = 0
  ): void {
    if (channelNumber < 0 || channelNumber >= this.numberOfChannels) {
      throw new IndexSizeError(
        `The channel number provided (${channelNumber}) is outside the range [0, ${this.numberOfChannels - 1}]`
      );
    }

    if (startInChannel < 0 || startInChannel >= this.length) {
      throw new IndexSizeError(
        `The startInChannel number provided (${startInChannel}) is outside the range [0, ${this.length - 1}]`
      );
    }

    this.buffer.copyFromChannel(destination, channelNumber, startInChannel);
  }

  public copyToChannel(
    source: Float32Array,
    channelNumber: number,
    startInChannel: number = 0
  ): void {
    if (channelNumber < 0 || channelNumber >= this.numberOfChannels) {
      throw new IndexSizeError(
        `The channel number provided (${channelNumber}) is outside the range [0, ${this.numberOfChannels - 1}]`
      );
    }

    if (startInChannel < 0 || startInChannel >= this.length) {
      throw new IndexSizeError(
        `The startInChannel number provided (${startInChannel}) is outside the range [0, ${this.length - 1}]`
      );
    }

    this.buffer.copyToChannel(source, channelNumber, startInChannel);
  }
}

File: packages/react-native-audio-api/common/cpp/audioapi/core/sources/AudioBuffer.cpp
#include <audioapi/core/sources/AudioBuffer.h>
#include <audioapi/utils/AudioArray.h>
#include <audioapi/utils/AudioBus.h>

#include <utility>

namespace audioapi {

AudioBuffer::AudioBuffer(
    int numberOfChannels,
    size_t length,
    float sampleRate) {
  bus_ = std::make_shared<AudioBus>(length, numberOfChannels, sampleRate);
}

AudioBuffer::AudioBuffer(std::shared_ptr<AudioBus> bus) {
  bus_ = std::move(bus);
}

size_t AudioBuffer::getLength() const {
  return bus_->getSize();
}

int AudioBuffer::getNumberOfChannels() const {
  return bus_->getNumberOfChannels();
}

float AudioBuffer::getSampleRate() const {
  return bus_->getSampleRate();
}

double AudioBuffer::getDuration() const {
  return static_cast<double>(getLength()) / getSampleRate();
}

float *AudioBuffer::getChannelData(int channel) const {
  return bus_->getChannel(channel)->getData();
}

void AudioBuffer::copyFromChannel(
    float *destination,
    size_t destinationLength,
    int channelNumber,
    size_t startInChannel) const {
  memcpy(
      destination,
      bus_->getChannel(channelNumber)->getData() + startInChannel,
      std::min(destinationLength, getLength() - startInChannel) *
          sizeof(float));
}

void AudioBuffer::copyToChannel(
    const float *source,
    size_t sourceLength,
    int channelNumber,
    size_t startInChannel) {
  memcpy(
      bus_->getChannel(channelNumber)->getData() + startInChannel,
      source,
      std::min(sourceLength, getLength() - startInChannel) * sizeof(float));
}

} // namespace audioapi
